karabo_bridge/simulation.py

```diff
@@ -309,41 +274,77 @@ def __init__(self, endpoint, ser='msgpack', protocol_version='2.2',
         self.generator = generate(det, nsources)
 
         self.zmq_context = zmq.Context()
-        self.server_socket = self.zmq_context.socket(zmq.REP)
+        if sock == 'REP':
+            self.server_socket = self.zmq_context.socket(zmq.REP)
+        elif sock == 'PUB':
+            self.server_socket = self.zmq_context.socket(zmq.PUB)
+        elif sock == 'PUSH':
+            self.server_socket = self.zmq_context.socket(zmq.PUSH)
+        else:
+            raise ValueError(f'Unsupported socket type: {sock}')
         self.server_socket.setsockopt(zmq.LINGER, 0)
+        self.server_socket.set_hwm(1)
         self.server_socket.bind(endpoint)
 
         self.stopper_r = self.zmq_context.socket(zmq.PAIR)
         self.stopper_r.bind('inproc://sim-server-stop')
         self.stopper_w = self.zmq_context.socket(zmq.PAIR)
         self.stopper_w.connect('inproc://sim-server-stop')
 
+        self.debug = debug
+
     def run(self):
         poller = zmq.Poller()
-        poller.register(self.server_socket, zmq.POLLIN)
+        poller.register(self.server_socket, zmq.POLLIN | zmq.POLLOUT)
         poller.register(self.stopper_r, zmq.POLLIN)
+
+        t_prev = time()
+        n = 0
+
         while True:
             events = dict(poller.poll())
-            if self.server_socket in events:
-                msg = self.server_socket.recv()
-                if msg == b'next':
-                    data, meta = next(self.generator)
-                    msg = serialize(data, meta,
-                                    protocol_version=self.protocol_version)
-                    self.server_socket.send_multipart(msg, copy=False)
-                else:
-                    print('Unrecognised request:', msg)
-            elif self.stopper_r in events:
+
+            if self.stopper_r in events:
                 self.stopper_r.recv()
                 break
+            if events[self.server_socket] is zmq.POLLIN:
+                msg = self.server_socket.recv()
+                if msg != b'next':
+                    print(f'Unrecognised request: {msg}')
+                    continue
+
+            data, meta = next(self.generator)
+            payload = serialize(data, meta,
+                                protocol_version=self.protocol_version)
+            try:
+                self.server_socket.send_multipart(payload, copy=False,
+                                                  flags=zmq.NOBLOCK)
+            except zmq.error.Again:
+                continue
```

Comment:
Hum, that would block if the client disconnect before sending the data. which happened in the tests... because I generate data (~.8s) between the poller and sending... 

That's fixed ^^'
thanks